### ISSUE TITLE
Return value of dispatch

### ISSUE THREAD

**Issue by gaearon** [gaearon/redux#61]
@vslinko 

What should `redux.dispatch()` return? Currently the code is

```js
    function dispatch(action) {
      return typeof action === 'function' ?
        action(dispatch, state) :
        dispatchSync(action);
    }
```

but I don't remember the circumstances when the return value of the callback form can be useful.

Because the callback form may schedule additional `dispatch`es, it seems that the calling code still doesn't have enough information. Can we just return whatever `action` was passed?

```js
    function dispatch(action) {
      if (typeof action === 'function') {
        action(dispatch, state);
      } else {
        dispatchSync(action);
      }
      return action;
    }
```

?

**Comment by vslinko** [comment]
My usage example

```js
export function runTransition(url) {
  return async perform => {
    const route = routing
      .filter(({route}) => route === url)
      .shift() || notFound

    const transition = transitions[route.transition]
    const componentKey = route.component

    const {title, status, ...props} = await perform(transition())

    perform(changePage(componentKey, props))

    return {
      url, title, status, componentKey, props
    }
  }
}

export function navigateTo(url) {
  return async perform => {
    progressIndicator(async () => {
      const {title} = await perform(runTransition(url))

      document.title = title
      history.pushState({url}, title, url)
    })
  }
}
```

I have `runTransition` action creator that fetching all data. This AC called on server side.
And I have `navigateTo` AC that calling `runTransition` and touching history API. This AC called on client side.

**Comment by vslinko** [comment]
Server side usage example:

```js

export default async function renderPage(url) {
  const dispatcher = createDispatcher()

  const {title, status} = await dispatcher.perform(runTransition(url))

  const atom = dispatcher.getAtom()

  const pageContent = React.renderToString(
    <Application dispatcher={dispatcher} />
  )

  return {
    status,
    body: appTemplate(title, pageContent, atom)
  }
}
```

**Comment by gaearon** [comment]
Thank you for the examples! Let's leave this open for now, I'll come back to it later..

**Comment by acdlite** [comment]
Returning a promise is the uber example. In all the Flux code I've ever written, I've only ever needed the return value for that reason.

**Comment by leoasis** [comment]
I'm not sure if the dispatch should return anything at all, the whole idea around the dispatcher is to do "fire and forget" and handle data that follows that action as app state, whenever it is received.

**Comment by acdlite** [comment]
I'm in favor of the the base `dispatch()` method return nothing. For those cases where you do need it to return a promise, that is easily implemented at the middleware level (refer to https://github.com/gaearon/redux/pull/63):

```js
function promiseMiddleware(next) {
  return action =>
    action && typeof action.then === 'function'
      ? action.then(next)
      : next(action);
}
```

**Comment by emmenko** [comment]
I also think the dispatcher should just "fire and forget"

**Comment by vslinko** [comment]
But how to implement action creators from my example if `dispatch` doesn't return anything?

**Comment by leoasis** [comment]
I don't know exactly what those `transition()` functions are doing and why you need to dispatch them. If `runTransition` is not being called as an AC, and only via `navigateTo` and `renderPage` then you can inject it the `dispatch` (or `perform`) method and make it a function that returns something and calls that dispatch as well. You have full control of that on your side, so no need to have the dispatcher return anything. Same thing can apply to those `transition()` functions

**Comment by acdlite** [comment]
@vslinko Couldn't you just get it from the atom? You have access to it right there a few lines down.

**Comment by acdlite** [comment]
*Middleware* can always return things, but the base dispatch method would not.

**Comment by vslinko** [comment]
> I don't know exactly what those transition() functions

Transition is action creator that fetches all data for page.
Transition should be action creator because it could change application state.
Transition should return value for `runTransition`
Simplified example:

```js
return function homepageTransition() {
  return async perform => {
    const apiResponse = (await homepageApiRequest())

    perform(mergeResources(collectResources(apiResponse)))

    return {
      status,
      title,
      queryParams,
      apiResponse,
    }
  }
}
```

`runTransition` is another action creator that mathes transition by url.
`runTransition` should be action creator because it could change application state.
`runTransition` should return value for `navigateTo`
Actial example:

```js
export function runTransition(url) {
  return async perform => {
    const route = routing
      .map(route => {
        return {...route, match: matchRoutePattern(route.route, url)}
      })
      .filter(({match}) => !!match)
      .shift()

    if (!route) {
      return perform(runNotFoundTransition(url))
    }

    const transition = transitions[route.transition]
    const componentKey = route.component

    const {title, status, ...props} = await perform(transition(route.match || {}))

    if (status === 404) {
      return perform(runNotFoundTransition(url))
    }

    if (status === 500) {
      return perform(runErrorTransition(url, props))
    }

    perform(changePage(componentKey, props))

    return {
      url, title, status
    }
  }
}
```


`navigateTo` is action creator too.
`navigateTo` should be action creator because it should perform another action creator.
Simplified example:

```js
export function navigateTo(url) {
  return async perform => {
    const {title} = await perform(runTransition(url))

    document.title = title
    history.pushState({url}, title, url)
  }
}
```

> @vslinko Couldn't you just get it from the atom? You have access to it right there a few lines down.

Only on server side. But how to get it in `navigateTo`?

**Comment by leoasis** [comment]
I think you can solve your problem if you manually call the other actionCreators by manually passing the `perform` function, which you already have in your root actionCreator. Then you can do:

instead of this:
```js
const {title} = await perform(runTransition(url))
```
You do this:
```js
const {title} = await runTransition(url)(perform)
```

And the same inside the `runTransition` action creator:

Instead of this:
```js
const {title, status, ...props} = await perform(transition(route.match || {}))
```
You do this:
```js
const {title, status, ...props} = await transition(route.match || {})(perform)
```

So conceptually you only call the dispatcher with one action creator from outside, and that action creator dispatches several actions, only the logic is spread in other functions that receive the dispatch function.

Another analysis that should be done is if you really need the transitions and the `runTransition` functions to actually be action creators (that is, actually called from the outside app directly), or just functions that can have the dispatch function injected.

**Comment by vslinko** [comment]
> I think you can solve your problem if you manually call the other actionCreators by manually passing the perform function, which you already have in your root actionCreator. Then you can do:

Yes, I know that.
Moreover I'm created PR that adds return to dispatcher because I don't want to pass `perform` and `state` into action creator by myself.
My original concerns described here https://github.com/gaearon/redux/issues/48#issuecomment-109578338

**Comment by leoasis** [comment]
Yes, you're right about that, you don't want to pass state just because. Anyway, I still think you shouldn't need to use the returned values from functions. For example in the `navigateTo` AC, you need the title and the url to set the document.title and the historyApi. If you managed those values in your app state (i.e. handled by a Store), you can then just model those inside your React application as components at the root level like `<DocumentTitle />` and `<CurrentUrl />`, that render nothing to the dom but modify something. Another option is to have an extra `subscribe` on the dispatcher and listen to state changes and directly set the title and url in that listener.

The server side part is trickier I think, since it's hard to know when there are no remaining actions to be dispatched if we cannot `await` on anything.

**Comment by acdlite** [comment]
You *can* await a promise if you use middleware. Just because the dispatcher doesn't return anything doesn't mean your middleware can't. See my example above https://github.com/gaearon/redux/issues/61#issuecomment-110393712

**Comment by leoasis** [comment]
Makes sense. But now I'm wondering if I didn't want to use promises but still wanted to know when the dispatch ends on the server side in order to send the request, how would I do that if i have ACs using the async function way?

**Comment by vslinko** [comment]
@leoasis if you don't use promises and async/await you could subscribe to dispatcher and wait for store changes. Your AC could set flag `completed` in store. That a reason I want to return value from AC.

@acdlite I don't mind about dispatcher "fire and forget" if middleware could change that behaviour. As I see middleware already in master so I'm satisfied.

**Comment by gaearon** [comment]
Is this solved my middleware in 0.10? Can we close?

**Comment by vslinko** [comment]
I didn't updated to 0.10 yet, so I don't know.
I think you could close this issue. I can (re)open issue If I'll have some questions.
